### DISCUSSION TITLE
Fallback to native implementations

### DISCUSSION THREAD

**Discussion by andreas-eriksson** [saucecontrol/PhotoSauce#154] [Q&A] [answered]
Would it be possible to automatically fallback to native codecs in the library on Windows computers that does not have the Media Feature Pack installed?

**Comment by saucecontrol** [comment]
I'm not sure I fully understand the ask.  You can basically achieve that today by registering the additional native codecs with `removeExisting: false`.

By default, all Microsoft-published WIC codecs will be registered first on Windows, before the `CodecManager.Configure` action runs.  With `removeExisting: false` on plugin registration, they will leave the WIC mapping in place and add their own mappings after, leaving WIC with higher priority.

**Reply by andreas-eriksson** [reply]
Does it always register the WIC codecs on Windows or does MagicScaler check if the Media Feature Pack is available?

I currently have the following check to see if it is installed so that I can then register native codecs.
```csharp
internal static bool MediaFeaturePackExists()
{
    using var hklm = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64);
    using var key = hklm.OpenSubKey(@"SOFTWARE\Microsoft\Windows\CurrentVersion\Setup\WindowsFeatures\WindowsMediaVersion");
    return key != null;
}
```

**Reply by andreas-eriksson** [reply]
I think I have misunderstood how things work. I had a browse through the source and WIC codecs are only added if they are available.
https://github.com/saucecontrol/PhotoSauce/blob/fa68cdc6bd6bd29bd7246ce2ee55784dca7c4142/src/MagicScaler/Codecs/CodecManager.cs#L291

This makes it easier for me as I can the do what you proposed and always add the native codecs as well.

**Reply by saucecontrol** [reply, marked as answer]
That's correct.  The WIC codecs are treated as plugins to the pipeline -- they're just registered automatically for convenience.  `UseWicCodecs` enumerates the codecs installed in WIC (through the registry), so there's no need to do your own check for installation of the media pack.

For reference, the codec collection passed to the `ConfigureCodecs` action will contain anything registered by `UseWicCodecs`.  You are free to examine that collection and make whatever changes you like before the collection is locked in by the pipeline after the action completes.
